### ISSUE TITLE
Bug: Icon on the desktop does not update after emptying the bin

### ISSUE THREAD

**Issue by yair100** [files-community/Files#9946]
**Steps to reproduce**
- Empty recycle bin
- See that the icon on the desktop still looks full

**Steps to fix**
- Update the icon to be empty

**Comment by hecksmosis** [comment]
Should this not be merged with #9195 ?

**Comment by QuaintMako** [comment]
A PR (#10113) that was implementing this issue has been made, but removed for limitation issues regarding the current implementation.

A few things to keep in mind for the next person implementing the feature:
- There are several scenarios where the icon needs to be reevaluated:
-- When files are deleted into the recycle bin
-- When files are being restored in the recycle bin
-- When files are being deleted from the recycle bin
-- When the recycle bin is entirely restored or empty

- The icons should be taken from Constants.ImageRes.
- The indexes for the Constants.ImageRes are namely public const int RecycleBin = 54 and public const int EmptyRecycleBin = 55.

**Comment by QuaintMako** [comment]
It seems there is an undocumented method from `Shell32.dll` called `SHUpdateRecycleBinIcon` that could update the icon of the RecycleBin. Might be useful for that issue.

**Comment by Lukiluc29** [comment]
@QuaintMako Personally in the last version of Files so the version 2.4.81, the icon of the Recycle Bin can is correctly updated on the desktop.

**Comment by NilayParikh33** [comment]
is this still open? I can try this one.

**Comment by Sergey842248** [comment]
Unfortunately, yes. It's still broken.

**Comment by yaira2** [comment]
> is this still open? I can try this one.

@NilayParikh33 please do 🙂 

I've had trouble fixing this.

**Comment by NilayParikh33** [comment]
Yes I can try but Actually I have to go through the code because I am new and I don't have understanding of code. So That's why.

**Comment by bogdan-patraucean** [comment]
I discovered this issue while playing with Files for like a few minutes, reported it to Yair on Discord and told me about what you already tried here and I managed to get a working solution that I tested with a console application.

What I did is that I added a few items to the recycle bin, closed File Explorer, added the Recyble Bin icon to the Desktop so I can confirm it's refreshed if I empty the bin from another process, other than explorer.exe, and then executed my console app that does this.

`NativeMethods.txt`
```txt
IShellItem
SHCreateItemFromParsingName
IFileOperation
SHEmptyRecycleBinW
CoInitializeEx
CoUninitialize

SHERB_*
```

`Program.cs`
```C#
using Windows.Win32;
using Windows.Win32.UI.Shell;
using Windows.Win32.System.Com;

Console.WriteLine("Hello, press any key when you want to empty the recycle bin!");

Console.ReadKey();

ForceRecycleBinEmptyIcon();

static bool EmptyRecycleBin()
{
    var hr = PInvoke.SHEmptyRecycleBin(new(), null, PInvoke.SHERB_NOCONFIRMATION | PInvoke.SHERB_NOPROGRESSUI | PInvoke.SHERB_NOSOUND);

    return hr == 0;
}

static void DeleteToRecycleBin(string path)
{
    PInvoke.CoInitializeEx(COINIT.COINIT_APARTMENTTHREADED);

    try
    {
        PInvoke.SHCreateItemFromParsingName(path, null, typeof(IShellItem).GUID, out var itemObj);

        var item = (IShellItem)itemObj;

        var fo = (IFileOperation?)Activator.CreateInstance(Type.GetTypeFromCLSID(new Guid("3AD05575-8857-4850-9277-11B85BDB8E09"))!);

        fo?.SetOperationFlags(FILEOPERATION_FLAGS.FOF_ALLOWUNDO | FILEOPERATION_FLAGS.FOF_NOCONFIRMATION | FILEOPERATION_FLAGS.FOF_SILENT);

        fo?.DeleteItem(item, null);

        fo?.PerformOperations();
    }
    finally
    {
        PInvoke.CoUninitialize();
    }
}

static void ForceRecycleBinEmptyIcon()
{
    {
        // Step 1: Empty (may not repaint)
        EmptyRecycleBin();
    }

    {
        // Step 2: Kick Explorer with a dummy file
        string temp = Path.Combine(Path.GetTempPath(), "RecycleBinKick.tmp");

        File.WriteAllText(temp, "dummy");

        DeleteToRecycleBin(temp);
    }

    {
        // Step 3: Empty again — now Explorer flips to empty
        EmptyRecycleBin();
    }
}
```

I used `IFileOperation` assuming this is what you use here since it's the modern approach and `SHEmptyRecycleBin` which is already used as well by Files. Basically what I do is:

- Empty the recycle bin with `SHEmptyRecycleBin` (will not repaint)
- Delete a temporary file that will redirect it to the Recycle Bin using `FOF_ALLOWUNDO` flag that will notify about the changes
- Empty the recyble bin again, this time Explorer will see the transition from full to empty and repaint

**Comment by yair100** [comment]
@bogdan-patraucean I tried implementing your suggestion, but it seems to have the same issue.

**Comment by 0x5bfa** [comment]
Did you try on a new STA thread? IIRC, we don't even need SHEmptyRecycleBin if we do shell file operation properly.

**Comment by bogdan-patraucean** [comment]
@yair100 last time I checked, it worked. Did you try the console app first? Would it help if I would test it again?

**Comment by myselfsiddharth** [comment]
I'd like to take a stab at this. I've read through the thread. aware of the failed PR #10113 and the four scenarios that need to be handled (delete, restore, permanent delete, full empty), using Constants.ImageRes indexes 54/55. Planning to look at what caused the previous attempt to fall short before writing any code. Happy to share my findings here first :)

**Comment by yair100** [comment]
@myselfsiddharth please keep us updated, thank you.

**Comment by myselfsiddharth** [comment]
@yair100 Update: opened a PR for this — https://github.com/files-community/Files/pull/18550

It calls `SHUpdateRecycleBinIcon` after all recycle bin mutations (empty, restore all, send to bin, restore items, permanent delete from bin) via a new `UpdateDesktopRecycleBinIcon()` method on `IStorageTrashBinService`.

Fixes #9946

**Comment by yair100** [comment]
I discovered that opening the delete confirmation dialog refreshes the Recycle Bin icon. With that in mind, I implemented a workaround in #18962 that appears to resolve this issue.
